lib/internal/test_runner/mock.js

```diff
@@ -0,0 +1,304 @@
+  }
+
+  get calls() {
+    return ArrayPrototypeSlice(this.#calls, 0);
+  }
+
+  callCount() {
+    return this.#calls.length;
+  }
+
+  mockImplementation(implementation) {
+    validateFunction(implementation, 'implementation');
+    this.#implementation = implementation;
+  }
+
+  mockImplementationOnce(implementation, onCall) {
+    validateFunction(implementation, 'implementation');
+    const nextCall = this.#calls.length;
+    const call = onCall ?? nextCall;
+    validateInteger(call, 'onCall', 0);
+
+    if (call < nextCall) {
+      // The call number has already passed.
+      return;
+    }
+
+    this.#mocks.set(call, implementation);
+  }
+
+  restore() {
+    const { descriptor, object, original, methodName } = this.#restore;
+
+    if (typeof methodName === 'string') {
+      // This is an object method spy.
+      ObjectDefineProperty(object, methodName, descriptor);
+    } else {
+      // This is a bare function spy. There isn't much to do here but make
+      // the mock call the original function.
+      this.#implementation = original;
+    }
+  }
+
+  trackCall(call) {
+    ArrayPrototypePush(this.#calls, call);
+  }
+
+  nextImpl() {
+    const nextCall = this.#calls.length;
+    const mock = this.#mocks.get(nextCall);
+    const impl = mock ?? this.#implementation;
+
+    if (nextCall + 1 === this.#times) {
+      this.restore();
+    }
+
+    this.#mocks.delete(nextCall);
+    return impl;
+  }
+}
+
+const { nextImpl, restore, trackCall } = MockFunctionContext.prototype;
+delete MockFunctionContext.prototype.trackCall;
+delete MockFunctionContext.prototype.nextImpl;
+
+class MockTracker {
+  #mocks;
+
+  constructor() {
+    this.#mocks = [];
+  }
+
+  fn(
+    original = function() {},
+    implementation = original,
+    options = kEmptyObject,
+  ) {
+    if (original !== null && typeof original === 'object') {
+      options = original;
+      original = function() {};
+      implementation = original;
+    } else if (implementation !== null && typeof implementation === 'object') {
+      options = implementation;
+      implementation = original;
+    }
+
+    validateFunction(original, 'original');
+    validateFunction(implementation, 'implementation');
+    validateObject(options, 'options');
+    const { times = Infinity } = options;
+    validateTimes(times, 'options.times');
+    const ctx = new MockFunctionContext(implementation, { original }, times);
+    return this.#setupMock(ctx, original);
+  }
+
+  method(
+    object,
+    methodName,
+    implementation = kDefaultFunction,
+    options = kEmptyObject,
+  ) {
+    validateObject(object, 'object');
+    validateStringOrSymbol(methodName, 'methodName');
+
+    if (implementation !== null && typeof implementation === 'object') {
+      options = implementation;
+      implementation = kDefaultFunction;
+    }
+
+    validateFunction(implementation, 'implementation');
+    validateObject(options, 'options');
+
+    const {
+      getter = false,
```

Comment:
can we expose `mock.getter(obj, 'key', impl)` as an alias for `mock.method(obj, 'key', impl, { getter: true })`?
same for `mock.setter(obj, 'key', impl)`

Reply:
Yea we can. I'd prefer adding that kind of sugar in follow up PRs though.

Reply:
sure thing